Fix loc legend validation

jpjepko opened the pull request:
## PR Summary
This pull request adds a check to the `loc` argument of `legend`. It checks that `loc` is a 2-tuple of floats or ints, and raises a `ValueError` otherwise. This addresses the issue #24605. The docstrings already mentioned the fact that `loc` is a pair of floats so we did not add any changes to any docstrings. We also added test cases for valid and invalid tuples.
## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

QuLogic commented on lib/matplotlib/legend.py:
```diff
@@ -472,6 +472,14 @@ def val_or_rc(val, rc_name):
                 loc = 'upper right'
         if isinstance(loc, str):
             loc = _api.check_getitem(self.codes, loc=loc)
+        if isinstance(loc, tuple):
```
Can be `elif`, since the `self.codes` above should always be valid.

QuLogic commented on lib/matplotlib/tests/test_legend.py:
```diff
@@ -1090,3 +1090,37 @@ def test_ncol_ncols(fig_test, fig_ref):
+def test_loc_short_tuple_exception():
+    # check that exception is raised if the loc arg
+    # of legend is not a 2-tuple of numbers
+    with pytest.raises(ValueError):
+        fig, ax = plt.subplots()
+        ax.legend(loc=(1.1, ))
```
Only the `legend` raises, so only it needs to be in the context manager. Also, please add a `match` so that it's certain which exception is being caught.

And I'm pretty sure all three of these invalid tests can be combined into one. Ditto for the valid ones.

tacaswell commented:
It looks like what happened is when you tried to push it failed, the message suggested you merge, you did so, and then the push worked?

What is going on is that by default git's push tries to keep you from accidentally make commits inaccessible so if a push _would_ orphan commits git fails and github helpfully suggest you merge the remote work with your and then push (both of which are good ideas).  However, in this case it is our goal to re-write history and we _want_ to make some commits disappear and replace them with new and improved commits so we need to tell git that we *really* mean it.  The incantation you want here is 

```
git push --force-with-leave YOUR_REPO fix-loc-legend-validation
```

which will tell git to do the push anyway, so long as you have at least had a chance to see the commits that you are going to be making unavailable (there is also `--force` that does in no questions asked but the extra checking in `--force-with-lease` saves me about once a month).

ksunden commented:
(just edited because there was a 1 character typo in there leave -> lease)

jpjepko commented:
I believe I had forgotten to point the upstream to the matplotlib repo
```
git remote add upstream https://github.com/matplotlib/matplotlib.git
```
Could that have been the culprit? In any case, would you prefer for me to open a new PR to keep the commit history clean?

QuLogic commented:
There's no need for that; rebase and drop the extra commits, then follow @tacaswell's instructions for force pushing the new ones.

jpjepko commented:
Thank you for your guidance. I cleaned up the commit history, rebased, and force pushed as per the updated [Development workflow](https://matplotlib.org/stable/devel/gitwash/development_workflow.html) in the dev docs. Are there any other changes you would like for me to incorporate?

jpjepko commented:
Hey @QuLogic, is there anything else that might need any changes? Thank you for your continued guidance.

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -472,6 +472,14 @@ def val_or_rc(val, rc_name):
         if isinstance(loc, str):
             loc = _api.check_getitem(self.codes, loc=loc)
+        elif isinstance(loc, tuple):
+            if len(loc) != 2:
+                raise ValueError(
+                    "loc must be string or pair of floats")
+            for loc_elt in loc:
+                # int can be converted to float, so ints are fine
+                if not isinstance(loc_elt, (float, int)):
+                    raise ValueError(f"{loc_elt} is not a valid value for loc")
```
```suggestion
            if len(loc) != 2 or not all(isinstance(e, Number) for e in loc):
                raise ValueError("loc must be string or pair of numbers")
        else:        
            raise ValueError("loc must be string or pair of numbers")
```
- simplify to check for [Number](https://docs.python.org/3/library/numbers.html#numbers.Number)
- merge the len and the number checks
- error out for other input types

QuLogic replied:
I think we want `numbers.Real`, as in:
https://github.com/matplotlib/matplotlib/blob/cc85fb66da61f85259d7d99f06faa6285f4ed225/lib/matplotlib/axes/_secondary_axes.py#L95-L96

jpjepko replied:
Just incorporated these changes in the original commit. Thanks!

tacaswell commented:
Sorry to be yet another person chiming in from the 🥜 gallery, but can you also add the repr of what the user _did_ pass to the error message?

tacaswell commented:
I'm going to milestone this as 3.8, if it is ready and merged it can definitely go in for 3.7, but I am trying to whittle down the milestone to things we would hold 3.8 over in the next 2 weeks!

QuLogic reviewed (changes requested):
It's not documented in the type, but it is documented in the description, that `loc` also accepts the numbers 0 through 10. You can add this (without magic numbers) by checking whether `loc in self.codes.items()`.

This should fix the failing tests.

tacaswell commented:
Pushed back to 3.8 as there are failing tests still.

jklymak commented:
@jpjepko are you hoping to come back to this?  Let us know if you need anything, but it looks like you are missing some valid loc values

rcomer commented:
Closing in favour of #25281.
